cachi2/core/package_managers/gomod.py

```diff
@@ -734,7 +740,6 @@ def _resolve_gomod(
     :raises PackageManagerError: if fetching dependencies fails
     """
     _protect_against_symlinks(app_dir)
-    modules_in_go_sum = _parse_go_sum(app_dir)
```

Comment:
Moving the parsing of the go.sum(s) to after downloading the dependencis is causing us a problem. It seems that Go has a tendency of adding missing hashes to `go.work.sum` when we invoke `go mod download`.

I did a few tests with our test [repo](https://github.com/ejegrova/cachi2-go-workspaces), and removing the hashes from a single dependency from both the `api/go.sum` and `gin/go.sum` folder resulted in that hash being instead added to `go.work.sum` :shrug:.

But keeping the parsing of the go.sum files before downloading seems to solve the issue (we might want to do further tests after we fix it).

Reply:
Moving the parsing of the go.sum files before downloading fixes the issue but creates different issues because we need to have workspace modules first to parse the sum files. The current approach of using cmd `go list -e -m -json` doesn't work well when vendoring flags are used. 

@brunoapimentel and @taylormadore  came with list of solutions:
- parse the `go.work` file manually to figure out where the workspaces are
- keep track of the `go.work.sum` file and, if it changed, fail the request (it [seems](https://go.dev/ref/mod#go-work-file) that Go adds missing checksums to the go.work.sum file instead of changing the workpaces go.sums)

Solutions that might not work:
- go list only if there are workspaces
    - vendoring for individual modules work even if there are workspaces
- handle vendoring first
    - calling go mod vendor might change the go.sum files
- handle the go list error individually (works, but I think it's a bad solution)
    - would affect all requests, regardless of workspaces being involved
    - would prevent users from fixing their vendoring with the `gomod-vendor` flag (as it would fail before that part of the code arrived) (edited) 
    
   @eskultety @ben-alkov @slimreaper35 any thoughts?

Reply:
Current preference is the first solution - parse `go.work` file, since the second one limits the user more.

Reply:
Just FYI, you can use `go work edit -json` to parse the file for you

Reply:
> Just FYI, you can use go work edit -json to parse the file for you

+1, but just beware of the toolchain version you use to parse the file, because Go < 1.21 won't parse the new toolchain version format from either `go.mod` or `go.work` (let alone the `toolchain` keyword):

```
cachi2-go-workspaces/go.work:1: invalid go version '1.21.0': must match format 1.23
cachi2-go-workspaces/go.work:3: unknown directive: toolchain
```
That's why I resorted in parsing go.mod manually. However, once we further tweak the toolchain selection (part of #500) we can use go's native mechanisms to parse either file and work solely with a dictionary representation.